homeassistant/config.py

```diff
@@ -669,6 +680,116 @@ def async_process_component_config(hass, config, domain):
     return None
 
 
+def check_ha_config_file(config_dir):
+    """Check if Home Assistant configuration file is valid."""
+    from unittest.mock import patch
+
+    all_errors = []
+    all_success = OrderedDict()
+
+    def _pack_error(package, component, config, message):
+        """Handle errors from packages: _log_pkg_error."""
+        print("XXXXXXXXXXXXXXXXXXXXX")
+        message = "Package {} setup failed. Component {} {}".format(
+            package, component, message)
+        domain = 'homeassistant.packages.{}.{}'.format(package, component)
+        pack_config = core_config[CONF_PACKAGES].get(package, config)
+        all_errors.append((message, domain, pack_config))
+
+    def _comp_error(ex, domain, config):
+        """Handle errors from components: async_log_exception."""
+        all_errors.append(
+            (_format_config_error(ex, domain, config), domain, config))
+
+    # Load configuration.yaml
+    try:
+        config_path = find_config_file(config_dir)
+        if not config_path:
+            return {}, [("File configuration.yaml not found.", None, None)]
+        config = load_yaml_config_file(config_path)
+    except HomeAssistantError as err:
+        return {}, [(str(err), None, None)]
+    finally:
+        clear_secret_cache()
+
+    # Extract and validate core [homeassistant] config
+    try:
+        core_config = config.pop(CONF_CORE, {})
+        core_config = CORE_CONFIG_SCHEMA(core_config)
+        all_success[CONF_CORE] = core_config
+    except vol.Invalid as err:
+        all_errors.append(err, CONF_CORE, core_config)
+        core_config = {}
+
+    # Merge packages
+    with patch('homeassistant.config._log_pkg_error', side_effect=_pack_error):
+        merge_packages_config(config, core_config.get(CONF_PACKAGES, {}))
```

Comment:
Patching the `_log_pkg_error` else I have to rewrite the complete method.

Another approach is to add the logging function to the funtion definition
```
def merge_packages_config(config, packages, _log_pkg_error=_log_pkg_error):
```

and then call it with `_log_pkg_error=_pack_error`

Reply:
passing method in is a 1000x better solution than patching.